openpype/hosts/substancepainter/api/pipeline.py

```diff
@@ -0,0 +1,347 @@
+            return
+
+        return filepath
+
+    def get_containers(self):
+
+        if not substance_painter.project.is_open():
+            return
+
+        metadata = substance_painter.project.Metadata(OPENPYPE_METADATA_KEY)
+        containers = metadata.get(OPENPYPE_METADATA_CONTAINERS_KEY)
+        if containers:
+            for key, container in containers.items():
+                container["objectName"] = key
+                yield container
+
+    @staticmethod
+    def create_context_node():
+        pass
+
+    def update_context_data(self, data, changes):
+        pass
+
+    def get_context_data(self):
+        pass
+
+    def _install_menu(self):
+        from PySide2 import QtWidgets
+        from openpype.tools.utils import host_tools
+
+        parent = substance_painter.ui.get_main_window()
+
+        menu = QtWidgets.QMenu("OpenPype")
+
+        action = menu.addAction("Load...")
+        action.triggered.connect(
+            lambda: host_tools.show_loader(parent=parent, use_context=True)
+        )
+
+        action = menu.addAction("Publish...")
+        action.triggered.connect(
+            lambda: host_tools.show_publisher(parent=parent)
+        )
+
+        action = menu.addAction("Manage...")
+        action.triggered.connect(
+            lambda: host_tools.show_scene_inventory(parent=parent)
+        )
+
+        action = menu.addAction("Library...")
+        action.triggered.connect(
+            lambda: host_tools.show_library_loader(parent=parent)
+        )
+
+        menu.addSeparator()
+        action = menu.addAction("Work Files...")
+        action.triggered.connect(
+            lambda: host_tools.show_workfiles(parent=parent)
+        )
+
+        substance_painter.ui.add_menu(menu)
+
+        def on_menu_destroyed():
+            self.menu = None
+
+        menu.destroyed.connect(on_menu_destroyed)
+
+        self.menu = menu
+
+    def _uninstall_menu(self):
+        if self.menu:
+            self.menu.destroy()
+            self.menu = None
+
+    def _register_callbacks(self):
+        # Prepare emit event callbacks
+        open_callback = partial(emit_event, "open")
+
+        # Connect to the Substance Painter events
+        dispatcher = substance_painter.event.DISPATCHER
+        for event, callback in [
+            (substance_painter.event.ProjectOpened, open_callback)
+        ]:
+            dispatcher.connect(event, callback)
+            # Keep a reference so we can deregister if needed
+            self.callbacks.append((event, callback))
+
+    def _deregister_callbacks(self):
+        for event, callback in self.callbacks:
+            substance_painter.event.DISPATCHER.disconnect(event, callback)
+        self.callbacks.clear()
+
+    def _install_shelves(self, project_settings):
+
+        shelves = project_settings["substancepainter"].get("shelves", {})
+        for name, path in shelves.items():
+            # TODO: Allow formatting with anatomy for the paths
```

Comment:
@antirotor Any pointers on how to best format a path using anatomy or alike during the host installation like this? :)

```python

from openpype.pipeline.template_data import get_template_data_with_names

anatomy = Anatomy(project_name)
anatomy_data = get_template_data_with_names(project_name, asset_name, task_name, system_settings)
path = anatomy.format(anatomy_data)
```
This doesn't seem like the right approach since it's not formatting against a template in anatomy but using a custom path. I need some way to specifically format the custom path. Some pseudocode:

```python
from openpype.pipeline.template_data import get_template_data_with_names

template_path = shelf["path"]
anatomy_data = get_template_data_with_names(project_name, asset_name, task_name, system_settings)
path = api.format(template_path, data=anatomy_data)
```

Reply:
Anatomy in fact is just templates, roots and attributes + helpers methods to work with them. So I would just use `anatomy.roots` in fill data, but don't use anatomy to format the template.

```python
from openpype.lib import StringTemplate
from openpype.pipeline.template_data import get_template_data_with_names

template_path = shelf["path"]
fill_data = get_template_data_with_names(project_name, asset_name, task_name, system_settings)
fill_data["root"] = anatomy.roots
path = StringTemplate.format_template(template_path, anatomy_data)
```